rbs/prism/AssertionsRewriterPrism.cc

```diff
@@ -1,16 +1,1140 @@
+            return prism.ConstantReadNode(original->name, original->base.location);
+        }
+
+        // Examples:
+        //   - Array[String] -> T::Array
+        //   - G1[Foo::Bar] -> Foo::Bar
+        //   - G1[MyModule::MyClass] -> MyModule::MyClass
+        case PM_CONSTANT_PATH_NODE: {
+            auto *original = down_cast<pm_constant_path_node_t>(node);
+            return prism.ConstantPathNode(original->base.location, deepCopyGenericTypeNode(parser, original->parent),
+                                          original->name);
+        }
+
+        // Examples:
+        //   - Pair[X] where X is a type parameter -> T.type_parameter(:X)
+        //   - The :X symbol needs to be copied with proper string allocation
+        case PM_SYMBOL_NODE: {
+            auto *original = down_cast<pm_symbol_node_t>(node);
+            auto symbolStr = parser.extractString(&original->unescaped);
+            auto loc = parser.translateLocation(original->base.location);
+            return prism.Symbol(loc, symbolStr);
+        }
+
+        default:
+            unreachable("deepCopyGenericTypeNode: unhandled node type {} - this node type appears in generic type "
+                        "instantiations but is not yet supported",
+                        PM_NODE_TYPE(node));
+    }
+}
+
+/**
+ * If `node` is a `.new` call and `type` is a generic type instantiation of the same class,
+ * replace the receiver of `.new` with the generic type.
+ *
+ * For example: `G1.new #: G1[Integer]` becomes `G1[Integer].new`.
+ */
+void maybeSupplyGenericTypeArguments(core::MutableContext ctx, parser::Prism::Parser &parser, pm_node_t **node,
+                                     pm_node_t **type) {
+    // We only rewrite `.new` calls
+    if (!PM_NODE_TYPE_P(*node, PM_CALL_NODE)) {
+        return;
+    }
+
+    auto *newCall = down_cast<pm_call_node_t>(*node);
+    auto methodName = parser.resolveConstant(newCall->name);
+
+    if (methodName != "new"sv) {
+        return;
+    }
+
+    // We only rewrite when casted to a generic type (syntheticSquareBrackets)
+    if (!PM_NODE_TYPE_P(*type, PM_CALL_NODE)) {
+        return;
+    }
+
+    auto *bracketCall = down_cast<pm_call_node_t>(*type);
+    auto bracketMethodName = parser.resolveConstant(bracketCall->name);
+
+    if (bracketMethodName != "<syntheticSquareBrackets>"sv) {
+        return;
+    }
+
+    // We only rewrite when the generic type is the same as the instantiated one
+    if (!sameConstant(parser, newCall->receiver, bracketCall->receiver)) {
+        return;
+    }
+
+    // We need to create a deep copy of the type node because it will be used in two places:
+    // 1. As the receiver of .new()
+    // 2. As the type argument to T.let()
+    newCall->receiver = deepCopyGenericTypeNode(parser, *type);
+}
+
+} // namespace
+
+/**
+ * Save the signature from the given block so it can be used to resolve the type parameters from the method signature.
```

Comment:
I think it's worth explaining how this context variable works, and what timing relationship it has to the other methods that rely on its side-effect